### ISSUE TITLE
FAILED VirtualProcessorSingleTest

### ISSUE THREAD

**Issue by johanandren** [akka/akka#20224]
https://jenkins.akka.io:8498/job/akka-multi-node-nightly/3685/consoleFull

```
[info] - stochastic_spec103_mustSignalOnMethodsSequentially *** FAILED ***
[info]   java.lang.AssertionError: Illegal concurrent access detected (entering critical section)! Thread[pool-184-thread-1,5,main] emited onNext(0) signal, before null finished its onSubscribe() signal.
[info]   at org.testng.Assert.fail(Assert.java:89)
[info]   at org.reactivestreams.tck.TestEnvironment.flop(TestEnvironment.java:117)
[info]   at org.reactivestreams.tck.PublisherVerification$5$1$1$ConcurrentAccessBarrier.enterSignal(PublisherVerification.java:270)
[info]   at org.reactivestreams.tck.PublisherVerification$5$1$1.onNext(PublisherVerification.java:303)
[info]   at akka.stream.impl.VirtualProcessor.rec$5(StreamLayout.scala:647)
[info]   at akka.stream.impl.VirtualProcessor.onNext(StreamLayout.scala:666)
[info]   at org.reactivestreams.example.unicast.AsyncIterablePublisher$SubscriptionImpl.doSend(AsyncIterablePublisher.java:161)
[info]   at org.reactivestreams.example.unicast.AsyncIterablePublisher$SubscriptionImpl.doRequest(AsyncIterablePublisher.java:91)
[info]   at org.reactivestreams.example.unicast.AsyncIterablePublisher$SubscriptionImpl.run(AsyncIterablePublisher.java:204)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
```

**Comment by drewhk** [comment]
https://jenkins.akka.io:8498/job/akka-multi-node/1474/consoleFull

**Comment by 2m** [comment]
https://jenkins.akka.io:8498/job/akka-multi-node-nightly/3770/consoleFull

**Comment by johanandren** [comment]
https://jenkins.akka.io:8498/job/pr-validator-per-commit-jenkins/6063/consoleFull

**Comment by johanandren** [comment]
https://jenkins.akka.io:8498/job/pr-validator-per-commit-jenkins/6136/consoleFull

**Comment by 2m** [comment]
https://jenkins.akka.io:8498/job/akka-multi-node/1541/consoleFull

**Comment by drewhk** [comment]
The failure is real. This is one of the issues with CAS loops :/ The generic pattern that causes these kind of errors is:
 1. attempt atomic state change by CAS loop
 2. after state change, fire side-effect

Our failure case is this:
 1. in the `subscribe` method of `VirtualProcessor`, when we have already have an upstream subscription, we finally CAS into the state `Both`.
 2. After the successful CAS, in `establishSubsciption` it calls `onSubscribe` on the downstream
 3. onSubscribe calls `request(1)` in the TCK. We have not yet left the `onSubscribe` method.
 4. `request(1)` is propagated to upstream (state is already `Both`)
 5. Upstream successfully delivers first element, `onNext(0)`, propagates to downstream because state is `Both`
 6. We have not yet exited onSubscribe:**BANG**

In other words, the issue is that the state `Both` is announced before `onSubscribe` returns, OTOH we need to do a CAS first to ensure that we have the right to call `onSubscribe` in the first place.

I don't yet know how to fix this properly. @rkuhn any ideas?

**Comment by viktorklang** [comment]
@drewhk introduce an intermediate state where request(n) requests are buffered? And then use getAndSet(0) to clear that out when transitioning to the final stage?

**Comment by drewhk** [comment]
But then you will have a race again:
 - getAndSet out the buffer and start to deliver it
 - races with unbuffered onNexts

**Comment by viktorklang** [comment]
@drewhk Can't we just hold off sending demand upstream until initialization is done?

**Comment by drewhk** [comment]
That is what happens though, read the sequence I provided again.

**Comment by rkuhn** [comment]
Hmm, I haven’t swapped in the full code info from when I implemented it, but it might be necessary to store the knowledge that we have not yet left `onSubscribe`, perhaps in a state and perhaps (if necessary) in a ThreadLocal :-(

**Comment by drewhk** [comment]
Yeah, one obvious solution is two-phase state switch, forcing someone to spin :(

In general, the code guarantees proper, linearized state-machine transitions, but it does not guarantee the linearization of side-effects performed on state transitions.

**Comment by rkuhn** [comment]
Forcing to spin (i.e. a custom spin-lock) is a no-go, that will eventually explode in production. Here we only need to defer passing demand upstream until we exit a certain stack frame, this should not require spinning.

**Comment by drewhk** [comment]
You mean we should buffer up demand while onSubscribe has not exited, and then fire it off? That might work.

**Comment by drewhk** [comment]
What I am not sure yet if this race can happen in other state transitions though.

**Comment by rkuhn** [comment]
Can’t look right now, but this needs to be done wherever onX methods are called.

**Comment by viktorklang** [comment]
That was what I proposed also.

-- 
Cheers,
√

**Comment by drewhk** [comment]
Ok, I misunderstood what you mean by "until initializiation is done". Why this can work is that requests do not need ordering, so the following is not an issue:
 1. call onSubscribe
 2. buffer up requests
 3. onSubscribe returns
 4. CAS state that now ungates requests
 5. downstream request arrives
 6. unbuffer demand from step 2

In step 5-6 there is a reversal, but this particular signal is not sensitive to it.

**Comment by drewhk** [comment]
Might trip up the TCK though...

**Comment by drewhk** [comment]
Although it shouldn't because the processor is free to transform requests as it sees fit. Great.
